### PR TITLE
chore: remove transformers cap and raise training floor

### PR DESCRIPTION
**Pull request by jaideepr97**
removes the temporary cap placed on the transformers library and bumps training to >=v0.8.1

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Functional tests have been added, if necessary.
- [ ] E2E Workflow tests have been added, if necessary.

### CODE REVIEW

**Review by booxter — APPROVED** [review]

**Review by RobotSail — APPROVED** [review]

**Comment by booxter** [comment]
the transformers cap merged, again; please update the PR to remove the cap.

**Review by booxter — COMMENTED** [review]
Please include transformers cap removal.

**Review by booxter — CHANGES_REQUESTED** [review]
.

**Comment by jaideepr97** [comment]
@booxter updated and also removed the comment above it 
GH UI is slow for some reason and not showing that yet

**Comment by booxter** [comment]
More issues in training library with the transformers?..

```
  0%|          | 0/5 [00:00<?, ?it/s]/actions-runner/_work/instructlab/instructlab/venv/lib64/python3.11/site-packages/transformers/generation/configuration_utils.py:820: UserWarning: `return_dict_in_generate` is NOT set to `True`, but `output_logits` is. When `return_dict_in_generate` is not `True`, `output_logits` is ignored.
  warnings.warn(
'tuple' object has no attribute 'logits'
```

**Comment by RobotSail** [comment]
@booxter It looks like the issue is the new transformers version changed their API. You can see in the [current implementation](https://github.com/instructlab/instructlab/blob/12b12173abc57272e79b6558667bd9896a8328e5/src/instructlab/model/full_train.py#L223) of the full train pipeline, it's getting the output by calling `return_dict=True`, but that warning is expecting `return_dict_in_generate=True` (I'm guessing to be set in the initialization).

https://github.com/instructlab/instructlab/blob/12b12173abc57272e79b6558667bd9896a8328e5/src/instructlab/model/full_train.py#L223

So the solution here is to either change the logic to just read the model output as a tuple or rename the parameter.

Long-term, we need to just move this function to live in the training library instead so we aren't duplicating this logic.

**Comment by jaideepr97** [comment]
hmm @RobotSail 
simply renaming the parameter to `return_dict_in_generation` didn't seem to work

File: src/instructlab/train/linux_train.py [hunk 1]
```diff
@@ -241,12 +242,15 @@ def linux_train(
     stopping_criteria = StoppingCriteriaList(
         [StoppingCriteriaSub(stops=stop_words_ids, device=model.device)]
     )
+    print("after tokenizer call")
```
Comment (booxter) [inline src/instructlab/train/linux_train.py:245]: remove debug prints

File: src/instructlab/train/linux_train.py [hunk 2]
```diff
@@ -232,6 +232,7 @@ def linux_train(
 
     print("LINUX_TRAIN.PY: SANITY CHECKING THE BASE MODEL")
     stop_words = ["<|endoftext|>", "<|assistant|>"]
+
```
Comment (booxter) [inline src/instructlab/train/linux_train.py:235]: nit: avoid unnecessary changes like this, they complicate git conflict resolutions later.

File: tests/test_package.py [hunk 3]
```diff
@@ -35,10 +35,10 @@ def test_provides_extra():
     assert set(m.get_all("Provides-Extra")).issuperset(HW_EXTRAS)
 
 
-def test_require_no_url_req():
-    # PyPI does not accept packages with URL requirements
-    for req in iter_requirements():
-        assert req.url is None, req
+# def test_require_no_url_req():
```
Comment (booxter) [inline tests/test_package.py:38]: revert?

File: src/instructlab/model/full_train.py [hunk 4]
```diff
@@ -220,16 +299,34 @@ def train(train_args, device, optimize_memory):
             for k in batch:
                 batch[k] = batch[k].to(device=dev)
 
-            output = model(**batch, use_cache=False, return_dict=True)
-            loss = output.loss
+            output = model(**batch, use_cache=False, return_dict=False)
+            loss = None
+            if isinstance(output, tuple):
+                loss = output[0]
+                assert (
```
Comment (booxter) [inline src/instructlab/model/full_train.py:306]: we should avoid asserts in code (there are different reasons for this, but one is: asserts can be disabled by runtime); if these were added for debugging purposes, then just remove them; if we think this is a legit scenario, then convert to `raise`s or handle the error case otherwise.

File: src/instructlab/train/linux_train.py [hunk 5]
```diff
@@ -232,6 +232,7 @@ def linux_train(
 
     print("LINUX_TRAIN.PY: SANITY CHECKING THE BASE MODEL")
```
[the remark this answers is not in the document]
Comment (jaideepr97) [reply src/instructlab/train/linux_train.py:233]: yeah - i can revert the entire file, probably
even without the changes we know it wasn't passing so i guess it doesnt matter either way

File: src/instructlab/model/full_train.py [hunk 6]
```diff
@@ -121,8 +195,7 @@ def setup_model(
     if tokenizer.eos_token_id is not None and model.config.eos_token_id is None:
         model.config.eos_token_id = tokenizer.eos_token_id
 
-    model = utils.convert_loss_to_reduce_sum(model)
-    model = utils.add_noisy_embeddings(model, noise_alpha=None)
+    model = convert_loss_to_reduce_sum(model)
```
[the remark this answers is not in the document]
Comment (jaideepr97) [reply src/instructlab/model/full_train.py:198]: changes to this file came from @RobotSail but i think they were done because a fix was needed in how we handle the output from transformers
I think it may have just been done to avoid having to do another training release but i could be wrong

[the remark this answers is not in the document]
Comment (booxter) [reply src/instructlab/model/full_train.py:205]: Please document the reason for the copy in a comment. And if it's a matter of another release of training, then add a TODO to remove the copy. (If it's just a matter of another training release, maybe we should have just cut it instead?)

File: requirements.txt [hunk 7]
```diff
@@ -46,3 +44,4 @@ haystack-ai>=2.8
 docling-core[chunking]>=2.10.0
 docling>=2.18.0
 sentence-transformers>=3.0.0
+bitsandbytes
```
Comment (booxter) [inline requirements.txt]: I believe it's not needed, and we don't import from the library directly. If there's some problem with other dependencies not tracking their reliance on bitsandbytes, then we should add a comment describing which dependency fails, ideally with a link to a bug report where this is reported upstream.

File: requirements/cuda.txt [hunk 8]
```diff
@@ -1,2 +1,2 @@
 # Extra dependencies for NVIDIA CUDA
-instructlab-training[cuda]>=0.8.0
```
Comment (booxter) [inline requirements/cuda.txt:2]: Are these changes here and below needed? Seems wrong. Could you please explain or revert?

File: src/instructlab/configuration.py [hunk 9]
```diff
@@ -1617,14 +1614,8 @@ def map_train_to_library(ctx, params):
     if params["pipeline"] == "full":
         train_args.disable_flash_attn = True
 
-    student_model_arch = get_model_arch(pathlib.Path(params["model_path"]))
     if ctx.obj.config.general.use_legacy_tmpl:
         train_args.use_legacy_tmpl = True
-    else:
```
Comment (booxter) [inline src/instructlab/configuration.py:1623]: does this change mean that legacy tokenizer config won't be detected anymore?

File: src/instructlab/model/full_train.py [hunk 10]
```diff
@@ -220,16 +306,38 @@ def train(train_args, device, optimize_memory):
             for k in batch:
                 batch[k] = batch[k].to(device=dev)
 
-            output = model(**batch, use_cache=False, return_dict=True)
-            loss = output.loss
+            output = model(**batch, use_cache=False, return_dict=False)
+            loss = None
+            if isinstance(output, tuple):
+                loss = output[0]
+                if len(output[0].shape) != 0:
```
Comment (booxter) [inline src/instructlab/model/full_train.py:313]: nit: `if len(loss.shape) != 0`? (Or even `if loss.shape`?)

File: src/instructlab/model/full_train.py [hunk 11]
```diff
@@ -220,16 +306,38 @@ def train(train_args, device, optimize_memory):
...
+                    raise ValueError(
+                        "When output is a tuple, the loss should be the first element"
+                    )
+            else:
+                loss = output.loss
+            if loss is None:
+                raise ValueError(
+                    "Loss is None. Ensure the model's output contains a valid loss."
```
Comment (booxter) [inline src/instructlab/model/full_train.py:321]: These errors will bubble up to users through CLI error. I am not sure they have the context to interpret the errors meaningfully. As a user, how do I "ensure the model's output contains a valid loss"? Are you asking the user to check their model inputs / config perhaps? If so, this is what should be communicated.

File: src/instructlab/model/full_train.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (RobotSail) [reply src/instructlab/model/full_train.py:205]: @booxter I don't see why we need to justify not relying on a function from a different library with different goals. The original code is designed to run on GPUs and handles other scenarios which we don't care about. Here we are changing it so that it literally just calculates the loss correctly. 

There's no reason to rely on functionality from a different library altogether as that adds unnecessary complexity for no benefit.

File: src/instructlab/model/full_train.py (hunk 11 above)
Comment (booxter) [reply src/instructlab/model/full_train.py:321]: This applies to other errors here I think?

File: src/instructlab/configuration.py (hunk 9 above)
Comment (RobotSail) [reply src/instructlab/configuration.py:1623]: @booxter We don't need it anymore

File: src/instructlab/model/full_train.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (booxter) [reply src/instructlab/model/full_train.py:205]: @RobotSail so is this function not going to be pulled from training library anymore in the future? This is not what @jaideepr97 communicated ("I think it may have just been done to avoid having to do another training release") so can we please clarify which it is?

As I said, the code itself is over my head, so my comments are procedural. I'd like to make sure that owner(s) of this PR at least know what they merge is actually important / needed. AFAIU we just removed a number of spurious changes from this PR that were included for no clear reason (changes to requirements files), so I have to wonder if this is the final shape we'd like to see.

File: src/instructlab/model/full_train.py [hunk 12]
```diff
@@ -257,7 +376,7 @@ def train(train_args, device, optimize_memory):
             torch.mps.empty_cache()
 
         output_dir = (
-            Path(train_args.ckpt_output_dir) / "hf_format" / f"samples_{(epoch*8)}"
+            Path(train_args.ckpt_output_dir) / "hf_format" / f"samples_{(epoch * 8)}"
```
[the remark this answers is not in the document]
Comment (RobotSail) [reply src/instructlab/model/full_train.py:379]: It's a ruff formatting change, I don't see why it matters to roll this back

File: src/instructlab/model/full_train.py (hunk 11 above)
Comment (RobotSail) [reply src/instructlab/model/full_train.py:321]: @booxter This shouldn't be happening, so if this happens then we want users to report this to us. They are also welcome to look at the internals of the CLI and potentially contribute back. We shouldn't assume that our users are incapable of being technical.

File: src/instructlab/model/full_train.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (RobotSail) [reply src/instructlab/model/full_train.py:205]: @booxter 
> so is this function not going to be pulled from training library anymore in the future?

This code entirely needs to live in the training repo to begin with. But right now there is no reason for us to be pulling on functions from the training repo for simple things.

File: src/instructlab/configuration.py (hunk 9 above)
Comment (booxter) [reply src/instructlab/configuration.py:1623]: We don't need the **change** or we don't need **support for legacy** auto-detection? And is this change related to transformers cap removal / training dependency update?

File: src/instructlab/model/full_train.py (hunk 11 above)
Comment (booxter) [reply src/instructlab/model/full_train.py:321]: I see. So it's **always** a bug in the code and not the inputs from the user? If so, the suggestion to "ensure the model's output" seems misplaced and we may instead want to direct the user to report the issue. (And maybe dump some more info to include with the report?)

Comment (RobotSail) [reply src/instructlab/model/full_train.py:321]: It could be a few things, the model implementation could have changed or the transformers API itself could have changed (as in the case here). So when that happens we just have this as a safeguard

**Review by cdoern — APPROVED** [review]
